### PR TITLE
ability to set proxies and choose the one using wildcard match

### PR DESCRIPTION
**Pull request by yaroslavzlotchenko**

### CODE REVIEW

**Comment by blino** [comment]
Can't we just use libproxy for this with its builtin envvar plugin?
It is automatically handled by GIO's proxy resolver, with the giolibproxy module from glib-networking.
This is a lot of code that we could avoid in WPE.

**Comment by yaroslavzlotchenko** [comment]
hi Olivier!
Thank you for your suggestion.
we thoroughly investigated the suggested approach (GIO extention point).
after considering it, we think that unfortunately it seems too complicated to fit our needs:

- we need to have different proxy pattern sets for each WebPage resource
- several web pages might be loaded at the same time with different proxy sets
- the proxy sets should have possibility to be changed dynamically 

so we would strongly prefer to add this API.

Please share your thoughts about it.

**Comment by blino** [comment]
Ok, can't the SoupNetworkSession::setHTTPProxy() be reused instead?
It makes use of g_simple_proxy_resolver_new(), not sure how much it would cover of your implementation in gwildcardproxyresolver

(Note: I am not a reviewer, just a contributor)

**Comment by yaroslavzlotchenko** [comment]
simple proxy resolver allows to set proxy only for the schemas while we need to have an
ability to set proxies for Glob-style patterns match

File: Source/WebCore/platform/Proxy.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef Proxy_h
```
Comment (zdobersek) [inline Source/WebCore/platform/Proxy.h]: Needs the license header, along with copyright information.

File: Source/WebCore/platform/network/soup/gwildcardproxyresolver.c [hunk 2]
```diff
@@ -0,0 +1,329 @@
+#include <string.h>
```
Comment (zdobersek) [inline Source/WebCore/platform/network/soup/gwildcardproxyresolver.c]: Ditto for the license header and copyright information.

File: Source/WebKit2/NetworkProcess/NetworkProcess.h [hunk 3]
```diff
@@ -186,6 +187,8 @@ class NetworkProcess : public ChildProcess, private DownloadManager::Client {
 #if USE(SOUP)
     void setIgnoreTLSErrors(bool);
     void userPreferredLanguagesChanged(const Vector<String>&);
+    void setProxies(WebCore::SessionID,
```
Comment (zdobersek) [inline Source/WebKit2/NetworkProcess/NetworkProcess.h]: No need for laying out each parameter into a separate line.

**Comment by zdobersek** [comment]
Looks OK, apart from the missing license and copyright information. Add that in and it's good to merge.

File: Source/WebCore/platform/Proxy.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (yaroslavzlotchenko) [inline Source/WebCore/platform/Proxy.h]: Not sure about the licence header, please correct if something wrong

File: Source/WebCore/platform/Proxy.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
```
Comment (igor-borovkov) [inline Source/WebCore/platform/Proxy.h]: 3rd clause

File: Source/WebCore/platform/Proxy.h (hunk 4 above)
Comment (zdobersek) [inline Source/WebCore/platform/Proxy.h]: Looks fine.

**Pull request merged**